synapse/handlers/message.py

```diff
@@ -225,6 +243,97 @@ def get_joined_members(self, requester, room_id):
             for user_id, profile in iteritems(users_with_profile)
         }
 
+    def maybe_schedule_expiry(self, event):
+        """Schedule the expiry of an event if there's not already one scheduled,
+        or if the one running is for an event that will expire after the provided
+        timestamp.
+
+        Args:
+            event (EventBase): The event to schedule the expiry of.
+        """
+        expiry_ts = event.content.get(EventContentFields.SELF_DESTRUCT_AFTER)
```

Comment:
an `assert not hs.config.worker_app` might be good here, both as a sanity check and a form of documentation that this should run on the master.